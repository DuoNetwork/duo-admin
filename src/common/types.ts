import { ChildProcess } from 'child_process';
export * from '../../../duo-contract-wrapper/src/types';

export interface IBaseMarketData {
	source: string;
	base: string;
	quote: string;
	timestamp: number;
}

export interface ITrade extends IBaseMarketData {
	id: string;
	price: number;
	amount: number;
}

export interface IPriceFix extends IBaseMarketData {
	price: number;
	volume: number;
}

export interface IPrice extends IBaseMarketData {
	period: number;
	open: number;
	high: number;
	low: number;
	close: number;
	volume: number;
}

export interface IBaseEvent {
	contractAddress: string;
	timestamp: number;
	transactionHash: string;
	blockNumber: number;
}

export interface IAcceptedPrice extends IBaseEvent {
	price: number;
	navA: number;
	navB: number;
}

export interface ITotalSupply extends IBaseEvent {
	tokenA: number;
	tokenB: number;
}

export interface IConversion extends IBaseEvent {
	type: string;
	eth: number;
	tokenA: number;
	tokenB: number;
	fee: number;
	pending?: boolean;
	reverted?: boolean;
}

export interface IStatus {
	process: string;
	timestamp: number;
}

export interface IPriceStatus extends IStatus {
	price: number;
	volume: number;
}

export interface INodeStatus extends IStatus {
	block: number;
}

export interface IAccount {
	address: string;
	privateKey: string;
}

export interface IKey {
	publicKey: string;
	privateKey: string;
}

export interface ISqlAuth {
	host: string;
	user: string;
	password: string;
}

export interface IOption {
	forceREST: boolean;
	pair: string;
	sources: string[];
	exSources: string[];
	assets: string[];
	live: boolean;
	dbLive: boolean;
	server: boolean;
	dynamo: boolean;
	aws: boolean;
	gcp: boolean;
	azure: boolean;
	force: boolean;
	gasPrice: number;
	gasLimit: number;
	source: string;
	event: string;
	provider: string;
	period: number;
	base: string;
	quote: string;
	contractType: string;
	tenor: string;
}

export interface ISourceSettings {
	priceInversed: boolean;
	quoteInversed: boolean;
	separator: string;
	isLowercase: boolean;
	tradesInterval: number;
	supportWS: boolean;
	filterByTimestamp: boolean;
	wsLink: string;
}

export interface ISourceAsset {
	base: {
		[code: string]: boolean;
	};
	quote: {
		[code: string]: boolean;
	};
	mapping: string;
	name: string;
}
export interface ISource {
	assets: {
		[code: string]: ISourceAsset;
	};
	settings: ISourceSettings;
}

export interface IBitfinexRawTradeWS {
	timestamp: number;
	seq: string;
	id: number;
	price: string;
	amount: string;
}

export interface IBitstampRawTradeRest {
	date: number;
	tid: number;
	price: number;
	type: string;
	amount: number;
}

export interface IBitstampRawTradeWs {
	microtimestamp: string;
	amount: number;
	buy_order_id: number;
	sell_order_id: number;
	amount_str: string;
	price_str: string;
	timestamp: string;
	price: number;
	type: number;
	id: number;
}

export interface IBitfinexRawTradeRest {
	timestamp: number;
	tid: number;
	price: string;
	amount: string;
	exchange: string;
	type: string;
}

export interface IGeminiTradeRest {
	timestamp: number;
	timestampms: number;
	tid: number;
	price: string;
	amount: string;
	exchange: string;
	type: string;
}

export interface IGeminiTradeWs {
	type: string;
	eventid: number;
	socket_sequence: number;
	timestamp: number;
	timestampms: number;
	events: IGeminiTradeData[];
}

export interface IGeminiTradeData {
	type: string;
	tid: number;
	price: string;
	amount: string;
	makerSide: string;
}

export type IKrakenRawTrade = IKrakenRawTradeEntry & {
	last: string;
};

export interface IKrakenRawTradeEntry {
	[pair: string]: string[][];
}

export interface IGdaxRawTradeRest {
	error: string[];
	result: {
		[code: string]: Array<Array<string | number>>;
	};
}

export interface ISubProcess {
	source: string;
	instance: ChildProcess;
	lastFailTimestamp: number;
	failCount: number;
}
