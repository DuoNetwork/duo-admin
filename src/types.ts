export interface ITrade {
	source: string;
	id: string;
	price: number;
	amount: number;
	timestamp: number;
}

export interface IPrice {
	price: number;
	volume: number;
	timestamp: number;
}

export interface IPriceBar {
	source: string;
	date: string;
	hour: string;
	minute: number;
	open: number;
	high: number;
	low: number;
	close: number;
	volume: number;
	timestamp: number;
}

export interface IEvent {
	type: string;
	id: string;
	blockHash: string;
	blockNumber: number;
	transactionHash: string;
	logStatus: string;
	parameters: {[key: string]: string};
	timestamp: number;
}

export interface IAccount {
	address: string;
	privateKey: string;
}

export interface IKey {
	publicKey: string;
	privateKey: string;
}

export interface ISqlAuth {
	host: string;
	user: string;
	password: string;
}

export interface IOption {
	live: boolean;
	server: boolean;
	dynamo: boolean;
	aws: boolean;
	gcp: boolean;
	azure: boolean;
	force: boolean;
	gasPrice: number;
	gasLimit: number;
	eth: number;
	address: string;
	addr1: string;
	addr2: string;
	privateKey: string;
	price: number;
	source: string;
	pwd: string;
	event: string;
	provider: string;
	contractState: string;
	accountNum: number;
	saveAccount: boolean;
	from: string;
	to: string;
	value: number;
	index: number;
	total: number;
	minEther: number;
	alpha: number;
	amtA: number;
	amtB: number;
	numOfMinutes: number;
	numOfHours: number;
	key: string;
	endBlk: number;
}
