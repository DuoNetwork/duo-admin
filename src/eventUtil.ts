import { Contract, EventLog } from 'web3/types';
import * as CST from './constants';
import ContractUtil from './contractUtil';
import dynamoUtil from './database/dynamoUtil';
import { IEvent, IOption } from './types';
import util from './util';

class EventUtil {
	public parseEvent(eventLog: EventLog, timestamp: number): IEvent {
		const returnValue = eventLog.returnValues;
		const output = {
			type: eventLog.event,
			id: eventLog['id'],
			blockHash: eventLog.blockHash,
			blockNumber: eventLog.blockNumber,
			transactionHash: eventLog.transactionHash,
			logStatus: eventLog['type'],
			parameters: {},
			timestamp: timestamp
		};
		for (const key in returnValue)
			if (!util.isNumber(key)) output.parameters[key] = returnValue[key];

		return output;
	}

	public async pull(
		contract: Contract,
		start: number,
		end: number,
		event: string
	): Promise<EventLog[]> {
		util.log('current blk is ' + end + ' last blk is ' + start + ' do subscription');
		return new Promise<EventLog[]>((resolve, reject) =>
			contract.getPastEvents(
				event,
				{
					fromBlock: start,
					toBlock: end
				},
				(error, events) => {
					if (error) reject(error);
					else resolve(events);
				}
			)
		);
	}

	public async subscribe(contractUtil: ContractUtil, option: IOption) {
		util.log('subscribing to ' + option.event);

		if (option.source)
			if ([CST.EVENT_START_PRE_RESET, CST.EVENT_START_RESET].includes(option.event))
				setInterval(async () => {
					const sysState = await contractUtil.readSysStates();
					const state = sysState[0];
					util.log('current state is ' + state);

					if (option.event === CST.EVENT_START_PRE_RESET && state === CST.STATE_PRERESET)
						await contractUtil.triggerPreReset();
					else if (
						option.event === CST.EVENT_START_RESET &&
						[CST.STATE_UP_RESET, CST.STATE_DOWN_RESET, CST.STATE_PERIOD_RESET].includes(
							state
						)
					)
						await contractUtil.triggerReset();
				}, 15000);
			else {
				let startBlk = (await dynamoUtil.readLastBlock()) || CST.INCEPTION_BLK;
				// let startBlk = CST.INCEPTION_BLK;
				util.log('last block number: ' + startBlk);
				setInterval(async () => {
					const currentBlk = await contractUtil.web3.eth.getBlockNumber();
					const block = await contractUtil.web3.eth.getBlock(currentBlk);
					if (startBlk <= currentBlk) {
						const allEvents: IEvent[] = [];
						const promiseList = CST.OTHER_EVENTS.map(async event => {
							const eventLogs: EventLog[] = await this.pull(
								contractUtil.contract,
								startBlk,
								currentBlk,
								event
							);
							eventLogs.forEach(async el => {
								const timestamp = (await contractUtil.web3.eth.getBlock(
									el.blockNumber
								)).timestamp;
								allEvents.push(this.parseEvent(el, timestamp));
							});
						});

						Promise.all(promiseList)
							.then(async () => {
								await dynamoUtil.insertEventData(allEvents);
								await dynamoUtil.insertStatusData({
									[CST.DB_ST_BLOCK]: { N: currentBlk + '' },
									[CST.DB_ST_TS]: { N: block.timestamp + '' },
									[CST.DB_ST_SYSTIME]: { N: util.getNowTimestamp() + '' }
								});
							})
							.catch(err => util.log(err));
					}
					startBlk = currentBlk + 1;
				}, 15000);
			}
		else {
			util.log('starting listening ' + option.event);
			let tg: (r: any) => Promise<void> = () => Promise.resolve();

			if ([CST.EVENT_START_PRE_RESET, CST.EVENT_START_RESET].includes(option.event)) {
				const sysState = await contractUtil.readSysStates();
				const state = sysState[0];
				util.log('current state is ' + state);

				if (option.event === CST.EVENT_START_PRE_RESET) {
					tg = () => contractUtil.triggerPreReset();
					if (state === CST.STATE_PRERESET) await contractUtil.triggerPreReset();
				} else if (option.event === CST.EVENT_START_RESET) {
					tg = () => contractUtil.triggerReset(2);
					if (
						[CST.STATE_UP_RESET, CST.STATE_DOWN_RESET, CST.STATE_PERIOD_RESET].includes(
							state
						)
					)
						await contractUtil.triggerReset();
				}
			} else return Promise.resolve();

			contractUtil.contract.events[option.event]({}, async (error, evt) => {
				if (error) util.log(error);
				else {
					console.log(evt);
					await tg(evt);
				}
			});
		}
	}
}

const eventUtil = new EventUtil();
export default eventUtil;
