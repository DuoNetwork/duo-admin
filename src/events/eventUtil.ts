import Web3 from 'web3';
import * as CST from '../constants';
import contractUtil from '../contractUtil';
// const provider = 'https://mainnet.infura.io/Ky03pelFIxoZdAUsr82w';
// const provider = 'ws://kovan.infura.io/WSDscoNUvMiL1M7TvMNP';
const provider = 'ws://localhost:8546';
// const provider = 'wss://socket.etherscan.io/wshandler';
const web3 = new Web3(new Web3.providers.WebsocketProvider(provider));
const CustodianABI = require('../static/Custodian.json'); // Custodian Contract ABI
const custodianContract = new web3.eth.Contract(CustodianABI['abi'], CST.CUSTODIAN_ADDR);

let gasPrice: number = 5e9;
const gasLimitPreReset = 120000;
const gasLimitReset = 7880000;

export class EventUtil {
	subscribeToAcceptPrice() {
		console.log('starting listening acceptPrice event');
		custodianContract.events.AcceptPrice(
			{
				fromBlock: CST.INCEPTION_BLK
			},
			function(error, event) {
				if (error) {
					console.log(error);
				} else {
					console.log(event);
				}
			}
		);
	}

	async trigger(abi: object, input: any[], gasPrice: number, gasLimit: number) {
		const nonce = await web3.eth.getTransactionCount(CST.PF_ADDR);
		const command = contractUtil.generateTxString(abi, input);
		// sending out transaction
		web3.eth
			.sendSignedTransaction(
				'0x' +
					contractUtil.signTx(
						contractUtil.createTxCommand(
							nonce,
							gasPrice,
							gasLimit,
							CST.CUSTODIAN_ADDR,
							0,
							command
						),
						CST.PF_ADDR_PK
					)
			)
			.on('receipt', console.log);
	}

	async triggerReset() {
		const abi = {
			name: 'startReset',
			type: 'function',
			inputs: []
		};
		const input = [];
		const web3GasPrice = await contractUtil.getGasPrice();
		gasPrice = web3GasPrice || gasPrice;
		console.log('gasPrice price ' + gasPrice + ' gasLimit is ' + gasLimitReset);
		await this.trigger(abi, input, gasPrice, gasLimitReset);
	}

	async triggerPreReset() {
		const abi = {
			name: 'startPreReset',
			type: 'function',
			inputs: []
		};
		const input = [];
		const web3GasPrice = await contractUtil.getGasPrice();
		gasPrice = web3GasPrice || gasPrice;
		console.log('gasPrice price ' + gasPrice + ' gasLimit is ' + gasLimitPreReset);
		await this.trigger(abi, input, gasPrice, gasLimitPreReset);  // 120000 for lastOne; 30000 for else
	}

	async subscribePreReset() {
		const state = await contractUtil.read('state');
		console.log(state);
		if (state === CST.STATE_PRERESET) await this.triggerPreReset();
		console.log('starting listening preReset event');
		custodianContract.events.StartPreReset(
			{
				fromBlock: CST.INCEPTION_BLK
			},
			async (error, event) => {
				if (error) {
					console.log(error);
				} else {
					console.log(event);
					await this.triggerPreReset();
				}
			}
		);
	}

	async subscribeReset() {
		console.log('starting listening reset event');
		const state = await contractUtil.read('state');
		if (
			state === CST.STATE_UP_RESET ||
			state === CST.STATE_DOWN_RESET ||
			state === CST.STATE_PERIOD_RESET
		) {
			console.log('start triggering reset');
			await this.triggerReset();
		}
		custodianContract.events.StartReset(
			{
				fromBlock: CST.INCEPTION_BLK
			},
			async (error, event) => {
				if (error) {
					console.log(error);
				} else {
					console.log(event);
					await this.triggerReset();
				}
			}
		);
	}
}

const eventUtil = new EventUtil();
export default eventUtil;
