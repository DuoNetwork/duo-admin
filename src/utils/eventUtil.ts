import BaseWrapper from '../../../duo-contract-wrapper/src/BaseWrapper';
import BeethovenWapper from '../../../duo-contract-wrapper/src/BeethovenWapper';
import Web3Wrapper from '../../../duo-contract-wrapper/src/Web3Wrapper';
import * as CST from '../common/constants';
import { IEvent, IOption } from '../common/types';
import dynamoUtil from './dynamoUtil';
import util from './util';

class EventUtil {
	public async trigger(
		address: string,
		privateKey: string,
		beethovenWappers: BeethovenWapper[],
		option: IOption
	) {
		util.logInfo('subscribing to ' + option.event);
		if (![CST.EVENT_START_PRE_RESET, CST.EVENT_START_RESET].includes(option.event)) {
			util.logError('invalid event, exit');
			return Promise.resolve();
		}

		if (option.source)
			setInterval(async () => {
				const promiseList = beethovenWappers.map(async bw => {
					const sysState = await bw.getStates();
					const state = sysState.state;
					util.logDebug('current state is ' + state + ' for ' + bw.address);

					if (option.event === CST.EVENT_START_PRE_RESET && state === CST.CTD_PRERESET)
						await bw.triggerPreReset(address, privateKey);
					else if (option.event === CST.EVENT_START_RESET && state === CST.CTD_RESET)
						await bw.triggerReset(address, privateKey);

					dynamoUtil.insertHeartbeat();
				});
				await Promise.all(promiseList);
			}, 15000);
		else {
			for (const bw of beethovenWappers) {
				util.logInfo('starting listening ' + option.event + ' for ' + bw.address);
				let tg: () => Promise<any> = () => Promise.resolve();
				const sysState = await bw.getStates();
				const state = sysState.state;
				util.logInfo('current state is ' + state + ' for ' + bw.address);

				if (option.event === CST.EVENT_START_PRE_RESET) {
					tg = () => bw.triggerPreReset(address, privateKey);
					if (state === CST.CTD_PRERESET) await tg();
				} else if (option.event === CST.EVENT_START_RESET) {
					tg = () => bw.triggerReset(address, privateKey);
					if (state === CST.CTD_RESET) await tg();
				}

				bw.contract.events[option.event]({}, async (error, evt) => {
					if (error) util.logInfo(error);
					else {
						util.logInfo(evt);
						await tg();
					}
				});
			}

			setInterval(() => dynamoUtil.insertHeartbeat(), 30000);
		}
	}

	public async fetch(baseWrappers: BaseWrapper[], force: boolean) {
		util.logInfo('fetching events');

		const web3Wrapper = baseWrappers[0].web3Wrapper;

		let startBlk = force
			? web3Wrapper.inceptionBlockNumber
			: Math.max(await dynamoUtil.readLastBlock(), web3Wrapper.inceptionBlockNumber);
		util.logInfo('starting blk number: ' + startBlk);
		let isProcessing = false;
		const fetch = async () => {
			if (isProcessing) return;

			isProcessing = true;
			const blockTimestampMap: { [blk: number]: number } = {};
			const currentBlk = await web3Wrapper.getCurrentBlockNumber();
			while (startBlk <= currentBlk) {
				const allEvents: IEvent[] = [];
				const end = Math.min(startBlk + CST.EVENT_FETCH_BLOCK_INTERVAL, currentBlk);
				const promiseList: Array<Promise<any>> = [];
				baseWrappers.forEach(bw =>
					bw.events.forEach(event =>
						promiseList.push(Web3Wrapper.pullEvents(bw.contract, startBlk, end, event))
					)
				);

				const results = await Promise.all(promiseList);
				for (const result of results)
					for (const el of result) {
						const blockTimestamp =
							blockTimestampMap[el.blockNumber] ||
							(await web3Wrapper.getBlockTimestamp(el.blockNumber));
						blockTimestampMap[el.blockNumber] = blockTimestamp;
						allEvents.push(Web3Wrapper.parseEvent(el, blockTimestamp));
					}

				util.logInfo(
					'total ' + allEvents.length + ' events from block ' + startBlk + ' to ' + end
				);
				if (allEvents.length > 0) await dynamoUtil.insertEventsData(allEvents);
				await dynamoUtil.insertHeartbeat({
					[CST.DB_ST_BLOCK]: { N: end + '' }
				});
				startBlk = end + 1;
			}
			isProcessing = false;
		};
		fetch();
		setInterval(() => fetch(), CST.EVENT_FETCH_TIME_INVERVAL);
	}
}

const eventUtil = new EventUtil();
export default eventUtil;
