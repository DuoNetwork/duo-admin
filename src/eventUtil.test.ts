import { EventLog } from 'web3/types';
import eventUtil from './eventUtil';
const sampleEvent: EventLog[] = require('./samples/eventsOther.json');

test('parseEvent', () =>
	sampleEvent.forEach(e => expect(eventUtil.parseEvent(e, 1234567890)).toMatchSnapshot()));
