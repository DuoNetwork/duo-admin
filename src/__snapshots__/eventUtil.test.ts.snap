// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseEvent 1`] = `
Object {
  "blockHash": "0xb236fabff690963c1e38ba2bc0ed8fff6f24d2c2a06d3437e587858a343e1562",
  "blockNumber": 7584187,
  "id": "log_fdade6f4",
  "logStatus": "mined",
  "parameters": Object {
    "ethAmtInWei": "94090000000000000",
    "redeemedTokenAInWei": "12472699044214333750",
    "redeemedTokenBInWei": "12472699044214333750",
    "sender": "0x00D8d0660b243452fC2f996A892D3083A903576F",
    "totalSupplyA": "0",
    "totalSupplyB": "0",
  },
  "timestamp": 1234567890,
  "transactionHash": "0xab624432b3cc6b1d8ada365be7899617f5ffa095b0e68bc054fdc02e309dc76e",
  "type": "Redeem",
}
`;

exports[`parseEvent 2`] = `
Object {
  "blockHash": "0xc018d11ef3c8ecd69376c555cf52b3b8323e9f5793cd1a2653fe25b58cd3fbff",
  "blockNumber": 7583611,
  "id": "log_fbe8fbfd",
  "logStatus": "mined",
  "parameters": Object {
    "navAInWei": "1000002000000000000",
    "navBInWei": "1085147861640112792",
    "priceInWei": "175481866396422670000",
    "sender": "0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a",
    "timeInSecond": "1528174500",
  },
  "timestamp": 1234567890,
  "transactionHash": "0x657c4d19de5efd60265f03059507fc801dec7795da7f80f8f4a31813a3e8cdfb",
  "type": "AcceptPrice",
}
`;

exports[`parseEvent 3`] = `
Object {
  "blockHash": "0x2faf54a3e08816dc4d13d651a3aaedadc3860353797ac62bc2b52eaaf4f05852",
  "blockNumber": 7583652,
  "id": "log_5a9a1166",
  "logStatus": "mined",
  "parameters": Object {
    "index": "0",
    "priceInWei": "200032406268784400000",
    "sender": "0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a",
    "timeInSecond": "1528174800",
  },
  "timestamp": 1234567890,
  "transactionHash": "0x2c8f0f87de5b1a08b8725c1dd33ffd6847920dfd29c3c6767fb9fe025f678914",
  "type": "CommitPrice",
}
`;

exports[`parseEvent 4`] = `
Object {
  "blockHash": "0x27c822696c709975f9aa01270ca037c265afa5cfb4dcb62e8dd1d16605afcc49",
  "blockNumber": 7583579,
  "id": "log_1ab4180f",
  "logStatus": "mined",
  "parameters": Object {
    "navAInWei": "1000000000000000000",
    "navBInWei": "1000000000000000000",
  },
  "timestamp": 1234567890,
  "transactionHash": "0xb5f7ddfeb52fc8680d4c9f6e24739b091732219602cecc679da63af485824d7b",
  "type": "StartTrading",
}
`;

exports[`parseEvent 5`] = `
Object {
  "blockHash": "0x3aedb0c0fb29d50e590f062a3fd708e8e043cd916760b316b2d6c56e0fd07750",
  "blockNumber": 7584181,
  "id": "log_d796919c",
  "logStatus": "mined",
  "parameters": Object {
    "createdTokenAInWei": "12472699044214333750",
    "createdTokenBInWei": "12472699044214333750",
    "sender": "0x00D8d0660b243452fC2f996A892D3083A903576F",
    "totalSupplyA": "12472699044214333750",
    "totalSupplyB": "12472699044214333750",
  },
  "timestamp": 1234567890,
  "transactionHash": "0x863d368faffe95427d36a3fe59ef6e83fc93b8e106d28df8f0cff023eadf4cdf",
  "type": "Create",
}
`;
